src/prefect/new_task_runners.py

```diff
@@ -133,12 +165,120 @@ def submit(
 
+    def map(
+        self,
+        task: "Task",
+        parameters: Dict[str, Any],
+        wait_for: Optional[Iterable[PrefectFuture]] = None,
+    ) -> Iterable[PrefectConcurrentFuture]:
+        """
+        Submit multiple tasks to the task run engine running in separate threads.
+
+        Args:
+            task: The task to submit.
+            parameters: The parameters to use when running the task.
+            wait_for: A list of futures that the task depends on.
+
+        Returns:
+            An iterable of future objects that can be used to wait for the tasks to
+            complete and retrieve the results.
+        """
+        if not self._started or self._executor is None:
+            raise RuntimeError("Task runner is not started")
+
+        from prefect.utilities.engine import (
+            collect_task_run_inputs_sync,
+            resolve_inputs_sync,
+        )
+
+        # We need to resolve some futures to map over their data, collect the upstream
+        # links beforehand to retain relationship tracking.
+        task_inputs = {
+            k: collect_task_run_inputs_sync(v, max_depth=0)
+            for k, v in parameters.items()
+        }
+
+        # Resolve the top-level parameters in order to get mappable data of a known length.
+        # Nested parameters will be resolved in each mapped child where their relationships
+        # will also be tracked.
+        parameters = resolve_inputs_sync(parameters, max_depth=0)
+
+        # Ensure that any parameters in kwargs are expanded before this check
+        parameters = explode_variadic_parameter(task.fn, parameters)
+
+        iterable_parameters = {}
+        static_parameters = {}
+        annotated_parameters = {}
+        for key, val in parameters.items():
+            if isinstance(val, (allow_failure, quote)):
+                # Unwrap annotated parameters to determine if they are iterable
+                annotated_parameters[key] = val
+                val = val.unwrap()
+
+            if isinstance(val, unmapped):
+                static_parameters[key] = val.value
+            elif isiterable(val):
+                iterable_parameters[key] = list(val)
+            else:
+                static_parameters[key] = val
+
+        if not len(iterable_parameters):
+            raise MappingMissingIterable(
+                "No iterable parameters were received. Parameters for map must "
+                f"include at least one iterable. Parameters: {parameters}"
+            )
+
+        iterable_parameter_lengths = {
+            key: len(val) for key, val in iterable_parameters.items()
+        }
+        lengths = set(iterable_parameter_lengths.values())
+        if len(lengths) > 1:
+            raise MappingLengthMismatch(
+                "Received iterable parameters with different lengths. Parameters for map"
+                f" must all be the same length. Got lengths: {iterable_parameter_lengths}"
+            )
```

Comment:
This logic was originally in the engine, but I moved it out to the task runner because each engine instance only handles on run. This feels like the best place right now for this logic, but it may be needed in other task runner implementations. If this logic is needed by other task runners, it may need to move to the base class.

Reply:
Nah, I like this down at this layer even if we have to duplicate it